AuthSystem Combined for DEMO

HaiyiCai opened the pull request:
Here's the latest version for DEMO, will debug after review.

HaiyiCai committed:
AuthSystem #74: Updated all Auth files in this branch for implement.

HaiyiCai committed:
Added the forgotten file for set up.

HaiyiCai committed:
Merge branch 'main' into haiyi-ms4 for coding and test.

HaiyiCai committed:
AuthSystem updated so far for testing.

HaiyiCai committed:
12/3 Update on UserModel & AuthController: email verification for register and password reset.

HaiyiCai committed:
Changed import format and comment format for consistency and readability.

HaiyiCai committed:
Update all so far for test and debug.

HaiyiCai committed:
Merge branch 'main' of https://github.com/batterydied/Aquatica into haiyi-ms4

HaiyiCai committed:
authSystem #74: Fixed spelling issues and updated server.js for test.

HaiyiCai committed:
AuthSystem #74: updated all except for AuthPage.

HaiyiCai committed:
AuthSystem #74: Fixed codes received from Devin, Winifred (review) and Benson (Database debug).

HaiyiCai committed:
AuthSystem #74: Necessary AuthService added with possible solution to Token Expiry mentioned in AuthRoutes.

HaiyiCai committed:
AuthSystem #74: Changed method for consistent database sync.

HaiyiCai committed:
Merge branch 'main' of https://github.com/batterydied/Aquatica into haiyi-ms4

HaiyiCai committed:
AuthSystem #74: merged current main and set up auth routes in server.js.

HaiyiCai committed:
Merge branch 'main' of https://github.com/batterydied/Aquatica into haiyi-ms4

HaiyiCai committed:
AuthPage #68:merging Devin's authpage with responding.

HaiyiCai committed:
Updated meon

HaiyiCai committed:
delete weird stuff

wnauda commented on backend/controllers/AuthController.js:
```diff
@@ -1 +1,318 @@
+
+// Includes logic for managing user authentication, including:
+/* 1. Registering a new user:  Handles user registration with required fields validations & password security. */
+export async function registerUser(req, res) {  // TODO Change name as register()
```
This method looks good, works as intended in my testing.

wnauda commented on backend/controllers/AuthController.js:
```diff
@@ -1 +1,318 @@
+
+
+/* 3. Log in a user: Verifies credentials, generates authentication tokens, and responds with user details.*/
+export async function login(req, res) {
```
Works as intended in my testing, looks good

wnauda commented on backend/controllers/AuthController.js:
```diff
@@ -1 +1,318 @@
+}
+
+/* 4. Log out a user: Invalidates tokens to ensure users can securely log out. */
+export async function logout(req, res) {
```
working!

wnauda commented on backend/controllers/AuthController.js:
```diff
@@ -1 +1,318 @@
+    console.log("Mail options created.");
+
+    console.log("Sending email through transporter...");
+    await transporter.sendMail(mailOptions);
```
this call is erroring out:
Error during password reset request: Error: Invalid login: 535-5.7.8 Username and Password not accepted. For more information, go to
535 5.7.8  https://support.google.com/mail/?p=BadCredentials 6a1803df08f44-6d8da9fe7f5sm8021006d6.80 - gsmtp

wnauda commented on backend/middleware/AuthMiddleware.js:
```diff
@@ -1 +1,64 @@
+import UserModel from "../models/UserModel.js";
+
+// Validates user credentials by checking the authentication token or session associated with the incoming request.
+const authMiddleware = async (req, res, next) => {
```
this is working now!

wnauda commented on backend/middleware/RoleMiddleware.js:
```diff
@@ -0,0 +1,49 @@
+  - Checks the roles attribute in the user object attached by AuthMiddleware.js.
+  - Confirms the user has the required role(s) for the route.
+*/
+      const userRole = req.user.roles;      // Extract role from authenticated user.
```
"req.user" is the user's id not a User object. You need to call getUserById() before you can retrieve the roles

wnauda commented on frontend/main/app/AppController.js:
```diff
@@ -26,29 +31,40 @@ export class AppController {
+     this.#currentView = authService.isLoggedIn() ? this.#views.marketplace : this.#views.auth;
    }
 
-   async render(id=null) {
```
The id parameter is still required to navigate to specific product pages, that change should be reversed.

wnauda reviewed (changes requested):
A few changes are needed, but overall you're on the right track!

wnauda commented on frontend/main/app/AppController.js:
```diff
@@ -6,17 +6,22 @@ import { ProductPage } from '../components/ProductPage/ProductPage.js';
 import { ProfilePage } from '../components/ProfilePage/ProfilePage.js';
 import { SellProductsPage } from '../components/MarketplacePage/SellProductsPage.js';
+import { AuthPage } from '../components/AuthPage/AuthPage.js';
+import { authService } from '../services/AuthService.js'; // Handles user authentication state
```
this is not a valid import. AuthService.js does not provide an export named authService, only the class AuthService. you should create an instance of the AuthService class in AuthService.js and export it if you want to use it like this.

wnauda commented on frontend/main/components/AuthPage/AuthPage.js:
```diff
@@ -0,0 +1,162 @@
+*/
+
+import { BaseComponent } from '../../app/BaseComponent.js';
+import { authService } from '../../services/AuthService.js';
```
same problem with authService as in AppController.js

wnauda commented on frontend/main/components/AuthPage/AuthPage.js:
```diff
@@ -0,0 +1,162 @@
+        }
+
+        const { token, userId } = await response.json();
+        authService.login(token, userId);
```
When I tested the login, it was successful on the backend but this call is failing and showing a failure alert.

wnauda commented on frontend/main/app/AppController.js:
```diff
@@ -62,15 +78,22 @@ export class AppController {
+    }
+
+     // Restrict access to certain views based on authentication
+    if(!authService.isLoggedIn() && (viewName !== 'auth' || 'marketplace' )){
```
|| 'marketplace' doesn't do anything here, 'marketplace' will always evaluate to true.

wnauda commented on frontend/main/services/AuthService.js:
```diff
@@ -0,0 +1,52 @@
+        return !!this.user && !this.isTokenExpired(this.user.token); 
+    }
+
+    isTokenExpired() {
```
this needs to have a parameter called token

wnauda commented on backend/controllers/AuthController.js:
```diff
@@ -1 +1,318 @@
+
+    res.status(200).json({
+      message: " ><> Login successful. <><",
+      token,
```
add a userId field with the value validCredentials.userId here

wnauda committed:
fixed auth bugs
- add userId to login response
- add parameter to isTokenExpired
- create instance of AuthService

wnauda committed:
allow not logged in user to access marketplace and auth page but no other pages

HaiyiCai commented:
Thnak you so much for helping me debugging and fixing the issue, Winifred!! I really appreciate your effort and contribution to Auth Sytem, we can't do this without you.

wnauda committed:
Merge branch 'main' into haiyi-ms4

wnauda committed:
fix nav

batterydied commented:
Hello, Winifred.
Check discord for my messages
